toolchain: add support Intel oneApi toolchain based on llvm

nashif opened the pull request:
This adds preliminary support for the oneApi toolchain for X86. The
toolchain is available from here:

   https://software.intel.com/content/www/us/en/develop/tools/oneapi/base-toolkit/download.html

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
toolchain: add support Intel oneApi toolchain based on llvm

This adds preliminary support for the oneApi toolchain for X86. The
toolchain is available from here:

   https://software.intel.com/content/www/us/en/develop/tools/oneapi/base-toolkit/download.html

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
@xfl888 FYI

ceolin commented:
why not simply:
```
include(${ZEPHYR_BASE}/cmake/compiler/clang/compiler_flags.cmake)
```

It seems the changes in this file are the same in clang.

nashif committed:
cmake: bintool: add llvm abstraction

Based on abstraction for gcc with slight modifications.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
cmake: llvm: use lld

Use lld linker instead of ld.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
doc: add oneApi toolchain

Document how to use oneApi toolkit with Zephyr.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif commented:
@tejlmand  can you please review this?

dcpleung commented on cmake/linker/lld/target.cmake:
```diff
@@ -0,0 +1,111 @@
+    # TODO: How would the linker script dependencies work for non-linker
+    # script generators.
+    message(STATUS "Warning; this generator is not well supported. The
+  Linker script may not be regenerated when it should.")
```
alignment?

tejlmand commented on cmake/bintools/oneApi/target.cmake:
```diff
@@ -0,0 +1,17 @@
+find_program(CMAKE_RANLIB  llvm-ranlib  ${find_program_clang_args}   )
+find_program(CMAKE_OBJCOPY objcopy      ${find_program_binutils_args})
+find_program(CMAKE_READELF readelf      ${find_program_binutils_args})
+
```
Interresting observation, when I run CMake multiple times I notice the following:
- First time: `CMAKE_STRIP` is undefined, thus running ninja result in: `&& --strip-all zephyr.elf -ozephyr.strip` being executed leading to following error:
  `/bin/sh: 1: --strip-all: not found`
- Second run: Re-running CMake using: `cmake .` gets me this as `CMAKE_STRIP` value: `CMAKE_STRIP-NOTFOUND`
- Third run: Re-running CMake once more: `cmake .` gets me this as `CMAKE_STRIP` value: `/usr/bin/strip`

nashif replied:
now finding CMAKE_STRIP which is found as `/opt/intel/oneapi/intelpython/latest/bin/llvm-strip`

tejlmand reviewed (changes requested):
Initial review and testing.

nashif commented on cmake/bintools/oneApi/target.cmake:
```diff
@@ -0,0 +1,17 @@
+
+find_program(CMAKE_AR      llvm-ar      ${find_program_clang_args}   )
+find_program(CMAKE_NM      llvm-nm      ${find_program_clang_args}   )
+find_program(CMAKE_OBJDUMP llvm-objdump ${find_program_clang_args}   )
```
there is `/opt/intel/oneapi/intelpython/latest/bin/llvm-objdump`

tejlmand replied:
Hmm, guess i'm missing something in my installation.
I disabled the Python, as it's almost 10Gb, and I just wanted the compiler.

Installing the `Intel distribution for Python` package helped.
10 Gb just for objdump and strip :laughing:

tejlmand replied:
would never have guessed those would be part of Python and not compiler if it were not for your link.

nashif commented on doc/getting_started/toolchain_3rd_party_x_compilers.rst:
```diff
@@ -50,6 +50,18 @@ GNU ARM Embedded
+#. Assuming the toolkit is installed in ``/opt/intel/oneApi``, set environment
+   using::
+
+        source  /opt/intel/oneapi/setvars.sh
```
The setvars.sh does lots of things, doing this manually and listing them all should not be something we want to tinker with IMO

nashif commented on cmake/bintools/oneApi/target.cmake:
```diff
@@ -0,0 +1,17 @@
+find_program(CMAKE_NM      llvm-nm      ${find_program_clang_args}   )
+find_program(CMAKE_OBJDUMP llvm-objdump ${find_program_clang_args}   )
+find_program(CMAKE_RANLIB  llvm-ranlib  ${find_program_clang_args}   )
+find_program(CMAKE_OBJCOPY objcopy      ${find_program_binutils_args})
```
we have it in /opt/intel/oneapi/intelpython/latest/bin/llvm-objcopy

nashif replied:
will change to llvm-objcopy, did not see any issues with gap-fill and x86 yet.

tejlmand replied:
I see this:
```
/opt/oneapi/intelpython/latest/bin/llvm-objcopy --gap-fill 0xff --output-target=binary --remove-section=.comment --remove-section=COMMON --remove-section=.eh_frame zephyr.elf zephyr.bin

/opt/oneapi/intelpython/latest/bin/llvm-objcopy: error: unknown argument '--gap-fill'
```

which is identical to earlier experience with vanilla llvm.
But gap fill is just a flash wearing improvement of filling gaps in flash with 0xFF, that is don't write anything compared to actually write 0x00.
I think it's acceptable to disable gap filling for toolchains not supporting this.

tejlmand replied:
Of course, one can build using:
```
 -DCONFIG_BUILD_NO_GAP_FILL=y
```
which points back to the fact we need to get proper toolchain / Kconfig structure in place for toolchain specific features.

nashif replied:
r

> I see this:
> 
> ```
> /opt/oneapi/intelpython/latest/bin/llvm-objcopy --gap-fill 0xff --output-target=binary --remove-section=.comment --remove-section=COMMON --remove-section=.eh_frame zephyr.elf zephyr.bin
> 
> /opt/oneapi/intelpython/latest/bin/llvm-objcopy: error: unknown argument '--gap-fill'
> ```
> 
> which is identical to earlier experience with vanilla llvm.

where do you see this? build what code and what platform?

tejlmand replied:
This happens when building for `up_squared_32`:
```
$ cmake -GNinja -DBOARD=up_squared_32 -Bbuild samples/hello_world/
....
-- Generating done
-- Build files have been written to: /projects/github/zephyr_ws/zephyr/build
$ ninja -Cbuild
ninja: Entering directory `build'
...
/opt/oneapi/intelpython/latest/bin/llvm-objcopy: error: unknown argument '--gap-fill'
ninja: build stopped: subcommand failed.
```

Build is working when using `qemu_x86`.

nashif replied:
aha, was testing with qemu. How do you disable this per toolchain and avoid setting CONFIG_BUILD_NO_GAP_FILL for each board?

tejlmand replied:
> aha, was testing with qemu. How do you disable this per toolchain and avoid setting CONFIG_BUILD_NO_GAP_FILL for each board?

we don't at the moment :disappointed: 
Hence the reason for this comment earlier:
> which points back to the fact we need to get proper toolchain / Kconfig structure in place for toolchain specific features.

So we can take it in, and I'll look into cleanup of the Kconfig / Toolchain.

nashif replied:
ok, addressed in the boards for now.

tejlmand commented on cmake/linker/lld/target.cmake:
```diff
@@ -0,0 +1,111 @@
+  else()
+    # TODO: How would the linker script dependencies work for non-linker
+    # script generators.
+    message(STATUS "Warning; this generator is not well supported. The
```
Should this warning be a warning ?
```suggestion
    message(WARNING "This generator is not well supported. The
```

tejlmand commented:
When using `source <path>/<to>/oneApi/setvars.sh` the system Python settings get overruled with somithing that doesn't match my current installation, causing `menuconfig` to fail.

```
$ ninja -Cbuild menuconfig
ninja: Entering directory `build'
[0/1] cd /projects/github/ncs/zephyr/build/zephyr/kconfig && /opt/cmake-3.17.1-Linux-x86_64/bin/cmake -E env ZEPHYR_BASE=/projects/github/ncs/zephyr ZEPHYR_T...ephyr/dts/bindings /opt/oneapi/intelpython/latest/bin/python3.7 /projects/github/ncs/zephyr/scripts/kconfig/menuconfig.py /projects/github/ncs/zephyr/Kconfi
Traceback (most recent call last):
  File "/projects/github/ncs/zephyr/scripts/kconfig/menuconfig.py", line 194, in <module>
    import curses
  File "/opt/oneapi/intelpython/latest/lib/python3.7/curses/__init__.py", line 13, in <module>
    from _curses import *
ModuleNotFoundError: No module named '_curses'
```

Notice: **File "/opt/oneapi/intelpython/latest/lib/python3.7/curses/__init__.py"**.

Workaround, tell CMake to use specific Python, like:
```
cmake ....  -DPYTHON_PREFER=/usr/bin/python3.8 
```
and menuconfig works again.
Unsure what else could potentially be broken with the python from `oneApi` being in use.

nashif commented:
> When using `source <path>/<to>/oneApi/setvars.sh` the system Python settings get overruled with somithing that doesn't match my current installation, causing `menuconfig` to fail

yes i know, but this is going to be done independently, i.e. get away from sourcing the file and instead pulling only what we need. I will mention this in the docs.

nashif committed:
boards: x86: oneApi and LLVM objcopy do not support --gap-fill

Work around the fact that llvm objcopy does not support --gap-fill right
now. This should be done on the toolchain level at some point, it is
currently not possible however.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

ceolin commented on cmake/compiler/icx/generic.cmake:
```diff
@@ -0,0 +1,10 @@
+
+find_program(CMAKE_C_COMPILER icx ${find_program_icx_args})
+
+find_program(CMAKE_DTS_PREPROCESSOR gcc)
```
can't oneapi be used instead of host GCC ?

nashif replied:
good point, just tried, and it actually works with icx

nashif commented:
@tejlmand  I think I have addressed all the feedback
